server/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueue.java

```diff
@@ -0,0 +1,360 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import com.carrotsearch.randomizedtesting.generators.RandomNumbers;
+import org.apache.lucene.util.Counter;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPoolInfo;
+import org.elasticsearch.threadpool.ThreadPoolStats;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class DeterministicTaskQueue extends AbstractComponent {
+
+    private final List<Runnable> runnableTasks = new ArrayList<>();
+    private List<DeferredTask> deferredTasks = new ArrayList<>();
+    private long currentTimeMillis;
+    private long nextDeferredTaskExecutionTimeMillis = Long.MAX_VALUE;
+
+    public DeterministicTaskQueue(Settings settings) {
+        super(settings);
+    }
+
+    /**
+     * @return whether there are any runnable tasks.
+     */
+    boolean hasRunnableTasks() {
+        return runnableTasks.isEmpty() == false;
+    }
+
+    /**
+     * @return whether there are any deferred tasks, i.e. tasks that are scheduled for the future.
+     */
+    boolean hasDeferredTasks() {
+        return deferredTasks.isEmpty() == false;
+    }
+
+    /**
+     * @return the current (simulated) time, in milliseconds.
+     */
+    long getCurrentTimeMillis() {
+        return currentTimeMillis;
+    }
+
+    /**
+     * Runs the first runnable task.
+     */
+    void runNextTask() {
```

Comment:
should this method be public?

The scheduling methods are public as well as the advanceTime. I think ` hasRunnableTasks`, `hasDeferredTasks`, `getCurrentTimeMillis` and  `runNextTask` should be as well.